Add support for typed task Labels.

jorajeev opened the pull request:
[PR98](https://github.com/awslabs/shuttle/pull/98) added support for associated a single untyped `Tag` with each task and thread.  As we've gained experience with Tags, we've increasingly felt a need to have a mechanism that is both better typed, and allows more than one tag to be associated with tasks. This commit introduces `Labels`, which are inspired by `Extensions` in [the http crate](https://docs.rs/http/latest/http/struct.Extensions.html). Users can attach any set of Labels to a task or thread, with the only caveat being that there can be at most one Label for a given type T.  This is not too onerous a restriction, since one can use the common [newtype idiom](https://doc.rust-lang.org/rust-by-example/generics/new_types.html) to easily work around this.

For tracing, we also provide a newtype `TaskName` that can be converted to and from a `String`. If the `TaskName` label is set for a task, tracing output will show the `TaskName` (in addition to the `TaskId`) to make logs easier to read.

Since the current functionality provided by `Tag` is superseded by `Labels`, we also mark `Tag` as deprecated.  They will be removed in a future release.

<!-- Enter your PR description here -->

---

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

sarsko commented on src/current.rs:
```diff
@@ -34,23 +39,74 @@ pub fn clock_for(task_id: TaskId) -> VectorClock {
+
+/// Remove a label of the given type for the specified task, returning the old label for the type, if any
+#[inline]
+pub fn remove_label_for_task<T: Clone + std::fmt::Debug + Send + Sync + 'static>(task_id: TaskId) -> Option<T> {
```
Nit: I think these should have the `std::fmt::` part removed.

jorajeev replied:
Agreed, and fixed.

sarsko commented on src/runtime/task/labels.rs:
```diff
@@ -0,0 +1,330 @@
+/*
+** Code directly copied from https://github.com/hyperium/http/blob/master/src/extensions.rs
+** but renaming 'Extensions' to 'Labels'
+*/
```
Would be nice with a doc comment explaining what the code does.

jorajeev replied:
Added.

jamesbornholt replied:
nit: `//!`

sarsko commented on tests/basic/labels.rs:
```diff
@@ -0,0 +1,337 @@
+    fn event(&self, event: &Event<'_>) {
+        let metadata = event.metadata();
+        let target = metadata.target();
+        if target.contains("shuttle") && target.ends_with("::runtime::execution") {
```
Why is this no longer `target() ==` and why is it split up (why is it not `target.ends_with("shuttle::runtime::execution")`?)

jorajeev replied:
It provides a bit more flexibility, and doesn't make the check much weaker.

sarsko commented on tests/basic/labels.rs:
```diff
@@ -0,0 +1,337 @@
+// - changing child Labels doesn't affect parent Labels
+// - child Labels are independent of each other
+async fn spawn_tasks(counter: Arc<AtomicUsize>) -> HashSet<(usize, usize)> {
+    set_label_for_task(me(), Ident(1));
```
Not super important, but I find the `set_label_for_task(me()` slightly annoying (the annoying part is needing two imports/calls to do the most common operation for `Labels`). Not super fan of `set_label_for_current_task` either (though I think I prefer that over the `set me`. Maybe `set_label(label: Label)` ?

jorajeev replied:
Yeah, I didn't want to duplicate the function set with `{get,set}_label_for_task` and `{get,set}_label_for_current_task`.  One alternative I considered was to allow the argument to be `Option<TaskId>`, and making `None` always mean the current task.  I wasn't sure how readable that would be, and it's just as many characters as `me()`.

jamesbornholt commented on src/runtime/task/mod.rs:
````diff
@@ -38,6 +41,73 @@ use waker::make_waker;
+/// }, None);
+/// ```
+#[derive(Clone)]
+pub struct LabelFn(pub Arc<dyn Fn(&mut Labels) + Send + Sync + 'static>);
````
Instead of this, maybe we could add a new method to `std::thread::Builder` that sets the labels for the new task atomically with its creation? Oh, but there's no builders for async tasks. That's too bad.

jorajeev replied:
Yeah, exactly.  It's also easier to add this to existing code that typically just calls `thread::spawn(move || { .. })` or `tokio::spawn(async move { .. })`.

jamesbornholt commented on src/runtime/task/mod.rs:
```diff
@@ -38,6 +41,73 @@ use waker::make_waker;
+/// It's often useful to modify or add new Labels to T.  One approach is to put label changes
+/// at the beginning of the closure that is passed to `spawn`, but this approach has the drawback
+/// that the changes are applied only when T is first selected for execution, and the closure
+/// is invoked.  To overcome this drawback, we introduce the `LabelFn` label, which allows
+/// users to define a function that modifies the labels inherited from the parent when T is
+/// initially created.  
```
Maybe a little more detailed: `LabelFn` should be used as a label on the parent task. When a task with a `LabelFn` label spawns a new task, the wrapped function will be applied to the child's label set at spawn time.

Also, `LabelFn` is a bit too generic as a name; maybe `ChildLabelFn` or something.

jorajeev replied:
Renamed to `ChildLabelFn` and updated the comment.

jamesbornholt commented on src/runtime/task/mod.rs:
```diff
@@ -464,14 +539,13 @@ pub(crate) struct ParkState {
 pub struct TaskId(pub(super) usize);
 
 impl Debug for TaskId {
+    // If the `TaskName` label is set, use that when generating the Debug string
```
Might be nice to make `std::thread::Builder` set the `TaskName` label on a new thread if it has a `.name()` set?

jorajeev replied:
Think I'm already doing that: `Builder::spawn` calls `thread::spawn_named` which calls `ExecutionState::spawn_thread` which calls `ExecutionState::set_labels_for_new_task` which sets the name if one was provided.

jorajeev replied:
I added a check to `tests::basic::thread::thread_builder_name` to check that the name provided to the builder is assigned when you call `get_name_for_task()`.

sarsko commented on tests/basic/labels.rs:
```diff
@@ -0,0 +1,337 @@
+async fn spawn_tasks(counter: Arc<AtomicUsize>) -> HashSet<(usize, usize)> {
+    set_label_for_task(me(), Ident(1));
+    #[allow(clippy::type_complexity)]
+    let handles: Vec<future::JoinHandle<(usize, usize, Vec<future::JoinHandle<(usize, usize)>>)>> = (0..2)
```
This resolves with `Vec<_>`

sarsko replied:
https://github.com/awslabs/shuttle/commit/496b60e92566a8ae5ba265ca94a1b0e5d2b7f4a2

jorajeev replied:
Nice catch; thanks

sarsko commented on tests/basic/labels.rs:
```diff
@@ -0,0 +1,337 @@
+fn spawn_threads(counter: Arc<AtomicUsize>) -> HashSet<(usize, usize)> {
+    set_label_for_task(me(), Ident(1));
+    #[allow(clippy::type_complexity)]
+    let handles: Vec<thread::JoinHandle<(usize, usize, Vec<thread::JoinHandle<(usize, usize)>>)>> = (0..2)
```
This resolves with `Vec<_>`

sarsko commented on tests/basic/labels.rs:
```diff
@@ -0,0 +1,337 @@
+                set_label_for_task(me(), Parent(parent));
+                set_label_for_task(me(), Ident(2 * parent + i));
+
+                let handles: Vec<thread::JoinHandle<(usize, usize)>> = (0..2)
```
This resolves with `Vec<_>`

sarsko commented:
(discussed with Rajeev, posting here to not forget it / have an entry where it may be discussed)

One thing which is lost from `Label`s vs `Tag`s is the ability to gain more information than a simple string in the tracing output from Shuttle. There is a question of whether there should be the possibility to have more information logged, for instance by exposing a way to provide a closure with a type signature ala `&[Label] -> Debug`.

sarsko commented:
`Send + Sync` bound can be removed: https://github.com/awslabs/shuttle/commit/982213842338aedf2ca93901eb01fb5e07c908ac

jorajeev commented:
Summary of changes since the first reviewed version:
* renamed `LabelFn` -> `ChildLabelFn`, and changed the function signature to `Fn(TaskId, &mut Labels)`, because there are situations where the `TaskId` is useful
* removed `Send + Sync` requirement for types usable as Labels

jamesbornholt commented on src/current.rs:
```diff
@@ -34,23 +40,74 @@ pub fn clock_for(task_id: TaskId) -> VectorClock {
+}
+
+/// Get a label of the given type for the specified task, if any
+#[inline]
```
Do these inline annotations end up having any effect?

jorajeev replied:
Yeah, I wondered about that, since the compiler can figure this out on its own anyway.  Removing.

jamesbornholt commented on src/current.rs:
```diff
@@ -34,23 +40,74 @@ pub fn clock_for(task_id: TaskId) -> VectorClock {
+
 /// Sets the `tag` field of the current task.
 /// Returns the `tag` which was there previously.
+#[allow(deprecated)]
```
Don't we also want these to be marked as `#[deprecated]`, since they're the public API we're about to remove?

jorajeev replied:
Agreed; done.

jamesbornholt commented on src/runtime/task/mod.rs:
```diff
@@ -38,6 +41,74 @@ use waker::make_waker;
+    }
+}
+
+/// By default, when a task or thread T is spawned, it inherits all labels from its parent.
```
nit: The docs style is to write a one-sentence summary (which shows up on index pages) and then the rest of the text as a second+ paragraph.

jamesbornholt commented on src/current.rs:
```diff
@@ -5,9 +5,15 @@
 //! example, a tool that wants to check linearizability might want access to a global timestamp for
 //! events, which the [`context_switches`] function provides.
```
Could we add a paragraph to this doc comment with a quick overview of labels and what they're useful for?

jorajeev replied:
Added.

jamesbornholt commented on src/runtime/task/mod.rs:
```diff
@@ -38,6 +41,74 @@ use waker::make_waker;
 
 pub(crate) const DEFAULT_INLINE_TASKS: usize = 16;
 
+/// To make debugging easier, if a task is assigned a `TaskName(s)` Label,
+/// Shuttle will display the String `s` in addition to the `TaskId` in debug output.
```
nit: same thing about docs style being one-line summary of "what the thing is", then more paragraphs

jorajeev replied:
Done

jamesbornholt reviewed (commented):
just docs nits
